File: tests/typescript/unit/transport-mcp/authenticate.test.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Transport MCP authentication unit tests.

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { authenticate, checkActiveAuthority, createMandateVerifier, extractBearer } from '../../../../packages/transport/mcp/ts/src/authenticate.js'

const revocations = {
  isRevoked: vi.fn(),
  markRevoked: vi.fn(),
  currentDelegationEpoch: vi.fn(),
  markDelegationEpoch: vi.fn(),
}

let issuerId = 0
const jwksByIssuer = new Map<string, JsonWebKey>()

beforeEach(() => {
  vi.stubGlobal('fetch', vi.fn(async (input: RequestInfo | URL) => {
    const url = typeof input === 'string' ? input : input instanceof URL ? input.href : input.url
    const issuer = url.replace(/\/\.well-known\/jwks\.json$/, '')
    const jwk = jwksByIssuer.get(issuer)
    if (!jwk) {
      return new Response(JSON.stringify({ keys: [] }), {
        status: 404,
        headers: { 'content-type': 'application/json' },
      })
    }
    return new Response(JSON.stringify({ keys: [jwk] }), {
      status: 200,
      headers: { 'content-type': 'application/json' },
    })
  }))
})

async function mintToken(
  claims: Record<string, unknown> = {},
  scopes = 'mcp:call',
): Promise<{ token: string; issuer: string; audience: string }> {
  const issuer = `https://issuer-${++issuerId}.example.com`
  const audience = 'resource://api'
  const key = await crypto.subtle.generateKey(
    { name: 'ECDSA', namedCurve: 'P-256' },
    true,
    ['sign', 'verify'],
  )
  const jwk = await crypto.subtle.exportKey('jwk', key.publicKey)
  Object.assign(jwk, { kid: 'kid-1', alg: 'ES256', use: 'sig' })
  jwksByIssuer.set(issuer, jwk)
  const header = base64url(JSON.stringify({ alg: 'ES256', kid: 'kid-1', typ: 'JWT' }))
  const payload = base64url(JSON.stringify({
    iss: issuer,
    aud: audience,
    sub: 'user-1',
    zone_id: 'zone-1',
    client_id: 'app-1',
    sid: 'sid-1',
    root_sid: 'root-1',
    use: 'resource',
    sub_type: 'user',
    jti: 'jti-1',
    scope: scopes,
    iat: Math.floor(Date.now() / 1000),
    exp: Math.floor(Date.now() / 1000) + 300,
    ...claims,
  }))
  const body = `${header}.${payload}`
  const signature = await crypto.subtle.sign(
    { name: 'ECDSA', hash: 'SHA-256' },
    key.privateKey,
    new TextEncoder().encode(body),
  )
  const token = `${body}.${base64url(new Uint8Array(signature))}`
  return { token, issuer, audience }
}

function base64url(value: string | Uint8Array): string {
  const bytes = typeof value === 'string' ? new TextEncoder().encode(value) : value
  let binary = ''
  for (const byte of bytes) binary += String.fromCharCode(byte)
  return btoa(binary).replaceAll('+', '-').replaceAll('/', '_').replace(/=+$/, '')
}

describe('transport-mcp authentication', () => {
  afterEach(() => {
    revocations.isRevoked.mockReset()
    revocations.markRevoked.mockReset()
    revocations.currentDelegationEpoch.mockReset()
    revocations.markDelegationEpoch.mockReset()
    jwksByIssuer.clear()
    vi.unstubAllGlobals()
  })

  it('extracts bearer tokens', () => {
    expect(extractBearer('Bearer token-1')).toBe('token-1')
    expect(extractBearer('bearer token-1')).toBe('token-1')
    expect(extractBearer('BEARER token-1')).toBe('token-1')
    expect(extractBearer(`Bearer\t token-1  `)).toBe('token-1')
    expect(extractBearer('Bearer   ')).toBeNull()
    expect(extractBearer(undefined)).toBeNull()
  })

  it('extracts bearer tokens from long whitespace headers in linear time', () => {
    expect(extractBearer(`Bearer ${' '.repeat(100_000)}token-1`)).toBe('token-1')
    expect(extractBearer(`Bearer ${' '.repeat(100_000)}`)).toBeNull()
  })

  it('rejects missing tokens without verification', async () => {
    const result = await authenticate('', {
      issuer: 'https://issuer.example.com',
      audience: 'resource://api',
      revocations,
    })
    expect(result).toMatchObject({
      ok: false,
      error: { code: 'missing_token', description: 'Missing bearer token' },
    })
    expect(revocations.isRevoked).not.toHaveBeenCalled()
  })

  it('rejects successful verification when no revocation store is supplied', async () => {
    const { token, issuer, audience } = await mintToken()

    await expect(authenticate(token, {
      issuer,
      audience,
      revocations: undefined as never,
    })).resolves.toMatchObject({
      ok: false,
      error: { code: 'invalid_token', description: 'Revocation store required' },
    })
  })

  it('returns the verified principal and checks session revocation', async () => {
    const { token, issuer, audience } = await mintToken({
      agent_session_id: 'agent-1',
      delegation_edge_id: 'edge-1',
      root_sid: 'root-1',
      delegation_chain: [{ application_id: 'app-parent' }],
      hop_count: 2,
      target: ['resource://api'],
    })
    revocations.isRevoked.mockResolvedValue(false)

    const result = await authenticate(token, {
      issuer,
      audience,
      zoneId: 'zone-1',
      requiredScopes: ['mcp:call'],
      requiredTargets: ['resource://api'],
      requireAgent: true,
      requireDelegation: true,
      requireChainContains: ['app-parent'],
      maxHopCount: 4,
      revocations,
    })

    expect(result).toMatchObject({
      ok: true,
      principal: {
        sub: 'user-1',
        zoneId: 'zone-1',
        clientId: 'app-1',
        sid: 'sid-1',
      rootSid: 'root-1',
      subType: 'user',
        scope: 'mcp:call',
        agentSessionId: 'agent-1',
        delegationEdgeId: 'edge-1',
        hopCount: 2,
      },
    })
    expect(revocations.isRevoked).toHaveBeenCalledWith('sid-1')
    expect(revocations.isRevoked).toHaveBeenCalledWith('root-1')
    expect(revocations.isRevoked).toHaveBeenCalledWith('agent-1')
    expect(revocations.isRevoked).toHaveBeenCalledWith('edge-1')
  })

  it('authenticates tokens from multiple issuers minted in one test', async () => {
    const first = await mintToken({ sid: 'sid-1' })
    const second = await mintToken({ sid: 'sid-2' })
    revocations.isRevoked.mockResolvedValue(false)

    await expect(authenticate(first.token, {
      issuer: first.issuer,
      audience: first.audience,
      revocations,
    })).resolves.toMatchObject({ ok: true, principal: { sid: 'sid-1' } })
    await expect(authenticate(second.token, {
      issuer: second.issuer,
      audience: second.audience,
      revocations,
    })).resolves.toMatchObject({ ok: true, principal: { sid: 'sid-2' } })
  })

  it('rejects revoked sessions after successful verification', async () => {
    const { token, issuer, audience } = await mintToken()
    revocations.isRevoked.mockResolvedValue(true)

    await expect(authenticate(token, {
      issuer,
      audience,
      revocations,
    })).resolves.toMatchObject({
      ok: false,
      error: { code: 'session_revoked', description: 'Session revoked' },
    })
  })

  it('rejects delegated tokens from stale graph epochs', async () => {
    const { token, issuer, audience } = await mintToken({
      delegation_edge_id: 'edge-1',
      delegation_graph_epoch: 7,
    })
    revocations.isRevoked.mockResolvedValue(false)
    revocations.currentDelegationEpoch.mockResolvedValue(8)

    await expect(authenticate(token, {
      issuer,
      audience,
      revocations,
    })).resolves.toMatchObject({
      ok: false,
      error: { code: 'delegation_stale', description: 'Delegation graph changed' },
    })
    expect(revocations.currentDelegationEpoch).toHaveBeenCalledWith('zone-1')
  })

  it.each([
    ['root authority', { root_sid: 'root-1' }, 'root-1'],
    ['agent session', { agent_session_id: 'agent-1' }, 'agent-1'],
    ['delegated permission', { delegation_edge_id: 'edge-1' }, 'edge-1'],
  ])('rejects %s revocation anchors after verification', async (_label, claims, revoked) => {
    const { token, issuer, audience } = await mintToken(claims)
    revocations.isRevoked.mockImplementation(async (anchor: string) => anchor === revoked)

    await expect(authenticate(token, {
      issuer,
      audience,
      revocations,
    })).resolves.toMatchObject({
      ok: false,
      error: { code: 'session_revoked', description: 'Session revoked' },
    })
  })

  it('supports active-execution checks after initial authentication', async () => {
    revocations.isRevoked.mockResolvedValue(false)
    await expect(checkActiveAuthority({
      sub: 'user-1',
      zoneId: 'zone-1',
      clientId: 'app-1',
      sid: 'sid-1',
      rootSid: 'root-1',
      use: 'resource',
      subType: 'user',
      jti: 'jti-1',
      issuedAt: 10,
      expiresAt: 20,
      scope: 'mcp:call',
    }, revocations, 21_000)).resolves.toMatchObject({
      code: 'invalid_token',
      description: 'Token expired during execution',
    })
  })

  it('rejects active-execution checks with missing sid', async () => {
    revocations.isRevoked.mockResolvedValue(false)

    await expect(checkActiveAuthority({
      sub: 'user-1',
      zoneId: 'zone-1',
      clientId: 'app-1',
      sid: '',
      rootSid: '',
      use: 'resource',
      subType: 'user',
      jti: 'jti-1',
      issuedAt: 10,
      expiresAt: 20,
      scope: 'mcp:call',
    }, revocations, 10_000)).resolves.toMatchObject({
      code: 'invalid_token',
      description: 'Token validation failed',
    })
  })

  it.each([
    ['missing required scope', { requiredScopes: ['admin:call'] }, {}, 'insufficient_scope', 'Missing scope: admin:call'],
    ['missing required target', { requiredTargets: ['resource://tools/calendar'] }, { target: ['resource://tools/files'] }, 'invalid_token', 'Token validation failed'],
    ['session mandate use', {}, { use: 'session' }, 'invalid_token', 'Token validation failed'],
    ['agent identity required', { requireAgent: true }, {}, 'agent_required', 'Agent identity required'],
    ['delegation required', { requireDelegation: true }, {}, 'delegation_required', 'Delegation required'],
    ['delegation chain mismatch', { requireChainContains: ['app-parent'] }, { delegation_chain: [{ application_id: 'app-child' }] }, 'chain_mismatch', 'Delegation chain missing application: app-parent'],
    ['hop count exceeded', { maxHopCount: 1 }, { hop_count: 2 }, 'hop_count_exceeded', 'Hop count exceeded'],
    ['invalid zone', { zoneId: 'zone-2' }, {}, 'invalid_zone', 'Token zone validation failed'],
  ])('maps identity verification failure: %s', async (_label, deps, claims, code, description) => {
    const { token, issuer, audience } = await mintToken(claims)

    await expect(authenticate(token, {
      issuer,
      audience,
      revocations,
      ...deps,
    })).resolves.toMatchObject({
      ok: false,
      error: { code, description },
    })
  })

  it('maps malformed tokens to invalid_token', async () => {
    await expect(authenticate('invalid.jwt.token', {
      issuer: 'https://issuer.example.com',
      audience: 'resource://api',
      revocations,
    })).resolves.toMatchObject({
      ok: false,
      error: { code: 'invalid_token', description: 'Token validation failed' },
    })
  })

  it('reuses verifier defaults and supports route-level requirements', async () => {
    const { token, issuer, audience } = await mintToken({ target: ['resource://api'] })
    revocations.isRevoked.mockResolvedValue(false)
    const verifier = createMandateVerifier({ issuer, audience, revocations })

    await expect(verifier.authorization(`Bearer ${token}`, {
      requiredScopes: ['mcp:call'],
      requiredTargets: ['resource://api'],
    })).resolves.toMatchObject({ ok: true, principal: { sid: 'sid-1' } })

    await expect(verifier.require({ requiredScopes: ['admin:call'] }).authenticate(token)).resolves.toMatchObject({
      ok: false,
      error: {
        code: 'insufficient_scope',
        description: 'Missing scope: admin:call',
        hint: 'Request a mandate that includes every required scope for this route.',
      },
    })
  })

  it('warms either an injected JWKS cache or the default issuer cache', async () => {
    const cache = { warm: vi.fn(async () => undefined) }
    const cached = createMandateVerifier({
      issuer: 'https://issuer-cache.example.com',
      audience: 'resource://api',
      revocations,
      jwksCache: cache as never,
    })
    await cached.warmup()
    expect(cache.warm).toHaveBeenCalledWith('https://issuer-cache.example.com')

    const { issuer, audience } = await mintToken()
    const verifier = createMandateVerifier({ issuer, audience, revocations })
    await expect(verifier.warmup()).resolves.toBeUndefined()
    expect(fetch).toHaveBeenCalledWith(`${issuer}/.well-known/jwks.json`, expect.anything())
  })
})

File: tests/go/unit/transport/mcp/authenticate_test.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Transport MCP authentication tests for bearer parsing, JWT claims, and revocation.

package transportmcp_test

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"encoding/base64"
	"encoding/json"
	"math/big"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/garudex-labs/caracal/packages/identity/go"
	"github.com/garudex-labs/caracal/packages/revocation/go"
	transportmcp "github.com/garudex-labs/caracal/packages/transport/mcp/go"
	"github.com/golang-jwt/jwt/v5"
)

func TestExtractBearer(t *testing.T) {
	if got, ok := transportmcp.ExtractBearer("Bearer token-1"); !ok || got != "token-1" {
		t.Fatalf("expected bearer token, got %q ok=%v", got, ok)
	}
	if got, ok := transportmcp.ExtractBearer("bearer token-1"); !ok || got != "token-1" {
		t.Fatalf("expected lowercase bearer token, got %q ok=%v", got, ok)
	}
	if got, ok := transportmcp.ExtractBearer("BEARER token-1"); !ok || got != "token-1" {
		t.Fatalf("expected uppercase bearer token, got %q ok=%v", got, ok)
	}
	if _, ok := transportmcp.ExtractBearer("Bearer   "); ok {
		t.Fatal("expected blank bearer token to be rejected")
	}
}

func TestAuthenticateRejectsMissingToken(t *testing.T) {
	_, authErr := transportmcp.Authenticate("", transportmcp.Options{})
	if authErr == nil || authErr.Code != transportmcp.ErrMissingToken {
		t.Fatalf("expected missing token error, got %#v", authErr)
	}
}

func TestAuthenticateAcceptsVerifiedTokenAndChecksRevocation(t *testing.T) {
	token, issuer, closeServer := mintToken(t, jwt.MapClaims{
		"scope":                  "mcp:call",
		"sid":                    "sid-1",
		"root_sid":               "root-1",
		"agent_session_id":       "agent-1",
		"delegation_edge_id":     "edge-1",
		"delegation_chain":       []map[string]any{{"application_id": "app-parent"}},
		"hop_count":              2,
		"client_id":              "app-1",
		"source_session_id":      "agent-root",
		"target_session_id":      "agent-1",
		"delegation_path":        []string{"edge-root", "edge-1"},
		"delegation_graph_epoch": 7,
		"target":                 []string{"resource://api"},
	})
	defer closeServer()
	store := revocation.NewInMemoryStore(time.Hour)

	claims, authErr := transportmcp.Authenticate(token, transportmcp.Options{
		Issuer:               issuer,
		Audience:             "resource://api",
		ZoneID:               "zone-1",
		RequiredScopes:       []string{"mcp:call"},
		RequiredTargets:      []string{"resource://api"},
		RequireAgent:         true,
		RequireDelegation:    true,
		RequireChainContains: []string{"app-parent"},
		MaxHopCount:          3,
		Revocations:          store,
	})
	if authErr != nil {
		t.Fatalf("unexpected auth error: %#v", authErr)
	}
	if claims.Sub != "user-1" || claims.RootSid != "root-1" || claims.AgentSessionID != "agent-1" || claims.DelegationEdgeID != "edge-1" || claims.HopCount != 2 {
		t.Fatalf("unexpected claims: %#v", claims)
	}
}

func TestAuthenticateRejectsRevokedSession(t *testing.T) {
	token, issuer, closeServer := mintToken(t, jwt.MapClaims{"scope": "mcp:call", "sid": "sid-1"})
	defer closeServer()
	store := revocation.NewInMemoryStore(time.Hour)
	if err := store.MarkRevoked("sid-1", time.Hour); err != nil {
		t.Fatalf("mark revoked: %v", err)
	}

	_, authErr := transportmcp.Authenticate(token, transportmcp.Options{
		Issuer:      issuer,
		Audience:    "resource://api",
		Revocations: store,
	})
	if authErr == nil || authErr.Code != transportmcp.ErrSessionRevoked {
		t.Fatalf("expected session_revoked, got %#v", authErr)
	}
}

func TestAuthenticateRejectsRevokedAuthorityAnchors(t *testing.T) {
	tests := []struct {
		name    string
		claims  jwt.MapClaims
		revoked string
	}{
		{name: "root", claims: jwt.MapClaims{"root_sid": "root-1"}, revoked: "root-1"},
		{name: "agent", claims: jwt.MapClaims{"agent_session_id": "agent-1"}, revoked: "agent-1"},
		{name: "delegation", claims: jwt.MapClaims{"delegation_edge_id": "edge-1"}, revoked: "edge-1"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			token, issuer, closeServer := mintToken(t, tt.claims)
			defer closeServer()
			store := revocation.NewInMemoryStore(time.Hour)
			if err := store.MarkRevoked(tt.revoked, time.Hour); err != nil {
				t.Fatalf("mark revoked: %v", err)
			}
			_, authErr := transportmcp.Authenticate(token, transportmcp.Options{
				Issuer:      issuer,
				Audience:    "resource://api",
				Revocations: store,
			})
			if authErr == nil || authErr.Code != transportmcp.ErrSessionRevoked {
				t.Fatalf("expected session_revoked, got %#v", authErr)
			}
		})
	}
}

func TestCheckActiveAuthorityRejectsExpiredExecution(t *testing.T) {
	store := revocation.NewInMemoryStore(time.Hour)
	authErr := transportmcp.CheckActiveAuthority(identity.Claims{
		Sid:       "sid-1",
		ExpiresAt: time.Now().Add(-time.Second).Unix(),
	}, store, time.Now())
	if authErr == nil || authErr.Code != transportmcp.ErrInvalidToken {
		t.Fatalf("expected invalid_token, got %#v", authErr)
	}
}

func TestAuthenticateMapsIdentityErrors(t *testing.T) {
	tests := []struct {
		name   string
		opts   transportmcp.Options
		claims jwt.MapClaims
		code   transportmcp.ErrorCode
	}{
		{name: "scope", opts: transportmcp.Options{RequiredScopes: []string{"admin:call"}}, claims: jwt.MapClaims{"scope": "mcp:call"}, code: transportmcp.ErrInsufficientScope},
		{name: "target", opts: transportmcp.Options{RequiredTargets: []string{"resource://tools/calendar"}}, claims: jwt.MapClaims{"scope": "mcp:call", "target": []string{"resource://tools/files"}}, code: transportmcp.ErrInvalidToken},
		{name: "session mandate", opts: transportmcp.Options{}, claims: jwt.MapClaims{"scope": "mcp:call", "use": "session"}, code: transportmcp.ErrInvalidToken},
		{name: "zone", opts: transportmcp.Options{ZoneID: "zone-2"}, claims: jwt.MapClaims{"scope": "mcp:call"}, code: transportmcp.ErrInvalidZone},
		{name: "agent", opts: transportmcp.Options{RequireAgent: true}, claims: jwt.MapClaims{"scope": "mcp:call"}, code: transportmcp.ErrAgentRequired},
		{name: "delegation", opts: transportmcp.Options{RequireDelegation: true}, claims: jwt.MapClaims{"scope": "mcp:call"}, code: transportmcp.ErrDelegationRequired},
		{name: "chain", opts: transportmcp.Options{RequireChainContains: []string{"app-parent"}}, claims: jwt.MapClaims{"scope": "mcp:call", "delegation_chain": []map[string]any{{"application_id": "app-child"}}}, code: transportmcp.ErrChainMismatch},
		{name: "hop", opts: transportmcp.Options{MaxHopCount: 1}, claims: jwt.MapClaims{"scope": "mcp:call", "hop_count": 2}, code: transportmcp.ErrHopCountExceeded},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			token, issuer, closeServer := mintToken(t, tt.claims)
			defer closeServer()
			tt.opts.Issuer = issuer
			tt.opts.Audience = "resource://api"
			_, authErr := transportmcp.Authenticate(token, tt.opts)
			if authErr == nil || authErr.Code != tt.code {
				t.Fatalf("expected %s, got %#v", tt.code, authErr)
			}
		})
	}
}

func TestAuthenticateMapsInvalidToken(t *testing.T) {
	_, authErr := transportmcp.Authenticate("not-a-jwt", transportmcp.Options{Issuer: "https://issuer.example.com", Audience: "resource://api"})
	if authErr == nil || authErr.Code != transportmcp.ErrInvalidToken {
		t.Fatalf("expected invalid_token, got %#v", authErr)
	}
}

func mintToken(t *testing.T, claims jwt.MapClaims) (string, string, func()) {
	t.Helper()
	identity.ResetJWKSCache()
	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		t.Fatalf("generate key: %v", err)
	}
	jwk := publicJWK(key.PublicKey)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/.well-known/jwks.json" {
			http.NotFound(w, r)
			return
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"keys": []map[string]any{jwk}})
	}))
	now := time.Now()
	base := jwt.MapClaims{
		"iss":       server.URL,
		"aud":       "resource://api",
		"sub":       "user-1",
		"zone_id":   "zone-1",
		"client_id": "app-1",
		"sid":       "sid-1",
		"root_sid":  "root-1",
		"use":       "resource",
		"sub_type":  "user",
		"jti":       "jti-1",
		"scope":     "mcp:call",
		"iat":       now.Unix(),
		"exp":       now.Add(5 * time.Minute).Unix(),
	}
	for k, v := range claims {
		base[k] = v
	}
	token := jwt.NewWithClaims(jwt.SigningMethodES256, base)
	token.Header["kid"] = "kid-1"
	signed, err := token.SignedString(key)
	if err != nil {
		t.Fatalf("sign token: %v", err)
	}
	return signed, server.URL, server.Close
}

func publicJWK(key ecdsa.PublicKey) map[string]any {
	return map[string]any{
		"kty": "EC",
		"crv": "P-256",
		"kid": "kid-1",
		"use": "sig",
		"alg": "ES256",
		"x":   base64.RawURLEncoding.EncodeToString(paddedCoordinate(key.X)),
		"y":   base64.RawURLEncoding.EncodeToString(paddedCoordinate(key.Y)),
	}
}

func paddedCoordinate(v *big.Int) []byte {
	out := make([]byte, 32)
	b := v.Bytes()
	copy(out[32-len(b):], b)
	return out
}
